File: guides/micronaut-microservices-distributed-tracing-xray/bookrecommendation/java/src/main/java/example/micronaut/BookController.java
package example.micronaut;

import io.micronaut.http.annotation.Controller;
import io.micronaut.http.annotation.Get;
import reactor.core.publisher.Flux;
import org.reactivestreams.Publisher;

@Controller("/books")
public class BookController {

    private final BookCatalogueOperations bookCatalogueOperations;
    private final BookInventoryOperations bookInventoryOperations;

    public BookController(BookCatalogueOperations bookCatalogueOperations,
                          BookInventoryOperations bookInventoryOperations) {
        this.bookCatalogueOperations = bookCatalogueOperations;
        this.bookInventoryOperations = bookInventoryOperations;
    }

    @Get
    public Publisher<BookRecommendation> index() {
        return Flux.from(bookCatalogueOperations.findAll())
                .flatMap(b -> Flux.from(bookInventoryOperations.stock(b.isbn()))
                        .filter(Boolean::booleanValue)
                        .map(rsp -> b)
                ).map(book -> new BookRecommendation(book.name()));
    }
}

File: guides/micronaut-rabbitmq-rpc/bookrecommendation/python/src/example/micronaut/book_controller.py
from typing import Annotated

from jakarta.inject import Inject
from micronaut.http.annotation import Get
from org.reactivestreams import Publisher
from reactor.core.publisher import Flux

from .book_recommendation import BookRecommendation
from .catalogue_client import CatalogueClient
from .inventory_client import InventoryClient


catalogue_client: Annotated[CatalogueClient, Inject]  # <1>
inventory_client: Annotated[InventoryClient, Inject]  # <1>


@Get("/books")  # <2>
def index() -> Publisher[BookRecommendation]:
    return Flux.from_(catalogue_client.find_all(b"")).flatMap(
        lambda books: Flux.fromIterable(books)
    ).flatMap(
        lambda book: Flux.from_(inventory_client.stock(book.isbn))
        .filter(lambda in_stock: bool(in_stock))
        .map(lambda _: book)
    ).map(lambda book: BookRecommendation(book.name))
